### ISSUE TITLE
.NET 9 support

### ISSUE THREAD

**Issue by pdevito3** [efcore/EFCore.NamingConventions#295]
> Error NU1107 : Version conflict detected for Microsoft.EntityFrameworkCore.Relational. Install/reference Microsoft.EntityFrameworkCore.Relational 9.0.0 directly to project PROJECT to resolve this issue. 
 PROJECT -> Microsoft.EntityFrameworkCore.Design 9.0.0 -> Microsoft.EntityFrameworkCore.Relational (>= 9.0.0) 
 PROJECT -> EFCore.NamingConventions 8.0.1 -> Microsoft.EntityFrameworkCore.Relational (>= 8.0.0 && < 9.0.0).
 
 hey y'all, any thoughts on when we'd get a .NET 9 supported version live? even as a prerelease -- currently blocking my project upgrades

**Comment by onionhammer** [comment]
Ditto

**Comment by sbwalker** [comment]
Blocked by this as well when upgrading to official .NET 9 release. Why was this not a problem with .NET 9 RC2?

**Comment by EduardoBocada** [comment]
Same thing here.

**Comment by roji** [comment]
@atrauzzi and others, you can use your own build as a temporary measure to get unblocked - the main branch already targets EF 9.

**Comment by fergalmoran** [comment]
> Seems unwieldy given that dotnet doesn't offer any way to override nuget packages with local projects.  

I think if you take a dependency on a package like this then you need to accept that it will not instantly track the core updates.

**Comment by roji** [comment]
> Seems unwieldy given that dotnet doesn't offer any way to override nuget packages with local projects.

You should be able to simply comment out your `<PackageReference>`, replacing it with a `<PackageReference>`.

> It looks like I'm waiting on some other npgsql related dependencies to be updated as well.

Npgsql 9.0 and EFCore.PG 9.0 should be released very soon (i.e. in the next couple of days).

**Comment by Bohdandn** [comment]
@atrauzzi 
> Seems unwieldy given that dotnet doesn't offer any way to override nuget packages with local projects.

```
<PackageReference Include="EFCore.NamingConventions" Version="8.0.3">
  <ExcludeAssets>All</ExcludeAssets>
</PackageReference>
<Reference Include="EFCore.NamingConventions">
  <HintPath>.\EFCore.NamingConventions.dll</HintPath>
  <Private>true</Private>
</Reference>
```

**Comment by maurobernal** [comment]
if main work in Net 9.0, when publish the package in Nuget?

**Comment by ralmsdeveloper** [comment]
I know that some are super excited for the new version of the conversion, there is something you can do temporarily, and that way you can use EF Core 9, just for example you will use SnakeCase and LowerCase.

```csharp
public static class ConvertionExtensions
{
    private static CultureInfo _culture;

    public static void ToSnakeCaseNames(this ModelBuilder modelBuilder)
    {
        _culture = CultureInfo.InvariantCulture;

        SetNames(modelBuilder, NamingConvention.SnakeCase);
    }

    public static void ToLowerCaseNames(this ModelBuilder modelBuilder)
    {
        _culture = CultureInfo.InvariantCulture;

        SetNames(modelBuilder, NamingConvention.LowerCase);
    }

    private static string? NameRewriter(this string name, NamingConvention naming)
    {
        if (string.IsNullOrEmpty(name)) return name;

        return naming == NamingConvention.SnakeCase
            ? SnakeCaseNameRewriter(name)
            : LowerCaseNameRewriter(name);
    }

    private enum NamingConvention
    {
        SnakeCase,
        LowerCase,
    }

    private static void SetNames(ModelBuilder modelBuilder, NamingConvention naming)
    {
        _culture = CultureInfo.InvariantCulture;

        foreach (var entity in modelBuilder.Model.GetEntityTypes())
        {
            entity.SetViewName(entity.GetViewName()?.NameRewriter(naming));
            entity.SetSchema(entity.GetSchema()?.NameRewriter(naming));
            entity.SetTableName(entity.GetTableName()?.NameRewriter(naming));

            foreach (var property in entity!.GetProperties())
            {
                property.SetColumnName(property.GetColumnName()?.NameRewriter(naming));
            }

            foreach (var key in entity.GetKeys())
            {
                key.SetName(key.GetName()?.NameRewriter(naming));
            }

            foreach (var key in entity.GetForeignKeys())
            {
                key.SetConstraintName(key.GetConstraintName()?.NameRewriter(naming));
            }

            foreach (var index in entity.GetIndexes())
            {
                index.SetDatabaseName(index.GetDatabaseName()?.NameRewriter(naming));
            }
        }
    }

    private static string? LowerCaseNameRewriter(string name)
        => name.ToLower(_culture);

    // https://github.com/efcore/EFCore.NamingConventions/blob/main/EFCore.NamingConventions/Internal/SnakeCaseNameRewriter.cs
    private static string SnakeCaseNameRewriter(string name)
    {
        var builder = new StringBuilder(name.Length + Math.Min(2, name.Length / 5));
        var previousCategory = default(UnicodeCategory?);

        for (var currentIndex = 0; currentIndex < name.Length; currentIndex++)
        {
            var currentChar = name[currentIndex];
            if (currentChar == '_')
            {
                builder.Append('_');
                previousCategory = null;
                continue;
            }

            var currentCategory = char.GetUnicodeCategory(currentChar);
            switch (currentCategory)
            {
                case UnicodeCategory.UppercaseLetter:
                case UnicodeCategory.TitlecaseLetter:
                    if (previousCategory == UnicodeCategory.SpaceSeparator ||
                        previousCategory == UnicodeCategory.LowercaseLetter ||
                        previousCategory != UnicodeCategory.DecimalDigitNumber &&
                        previousCategory != null &&
                        currentIndex > 0 &&
                        currentIndex + 1 < name.Length &&
                        char.IsLower(name[currentIndex + 1]))
                    {
                        builder.Append('_');
                    }

                    currentChar = char.ToLower(currentChar, _culture);
                    break;

                case UnicodeCategory.LowercaseLetter:
                case UnicodeCategory.DecimalDigitNumber:
                    if (previousCategory == UnicodeCategory.SpaceSeparator)
                    {
                        builder.Append('_');
                    }
                    break;

                default:
                    if (previousCategory != null)
                    {
                        previousCategory = UnicodeCategory.SpaceSeparator;
                    }
                    continue;
            }

            builder.Append(currentChar);
            previousCategory = currentCategory;
        }

        return builder.ToString().ToLower(_culture);
    }
}
```
In your OnModelCreating add the call!
```csharp
protected override void OnModelCreating(ModelBuilder builder)
{ 
    builder.ToSnakeCaseNames();
   // or 
   //    builder.ToLowerCaseNames();
}
```

With this you should be able to update your application!

FYI: Remove the dependency on EFCore.NamingConventions temporarily.

**Comment by ericardezpsa** [comment]
> > Error NU1107 : Version conflict detected for Microsoft.EntityFrameworkCore.Relational. Install/reference Microsoft.EntityFrameworkCore.Relational 9.0.0 directly to project PROJECT to resolve this issue.
> > PROJECT -> Microsoft.EntityFrameworkCore.Design 9.0.0 -> Microsoft.EntityFrameworkCore.Relational (>= 9.0.0)
> > PROJECT -> EFCore.NamingConventions 8.0.1 -> Microsoft.EntityFrameworkCore.Relational (>= 8.0.0 && < 9.0.0).
> 
> hey y'all, any thoughts on when we'd get a .NET 9 supported version live? even as a prerelease -- currently blocking my project upgrades

Same here

**Comment by martinib77** [comment]
I'm having only as a warning, not an error

```
warning NU1608: Detected package version outside of dependency constraint: EFCore.NamingConventions 8.0.3 requires Microsoft.EntityFrameworkCore (>= 8.0.0 && < 9.0.0) but version Microsoft.EntityFrameworkCore 9.0.0 was resolved.
/Users/user/src/sisa2/backend/Sisa.Application.Data/Sisa.Application.Data.csproj : warning NU1608: Detected package version outside of dependency constraint: EFCore.NamingConventions 8.0.3 requires Microsoft.EntityFrameworkCore.Relational (>= 8.0.0 && < 9.0.0) but version Microsoft.EntityFrameworkCore.Relational 9.0.0 was resolved.
```

Any ideas why is that ?

**Comment by mhosman** [comment]
> I know that some are super excited for the new version of the conversion, there is something you can do temporarily, and that way you can use EF Core 9, just for example you will use SnakeCase and LowerCase.
> 
> ```cs
> public static class ConvertionExtensions
> {
>     private static CultureInfo _culture;
> 
>     public static void ToSnakeCaseNames(this ModelBuilder modelBuilder)
>     {
>         _culture = CultureInfo.InvariantCulture;
> 
>         SetNames(modelBuilder, NamingConvention.SnakeCase);
>     }
> 
>     public static void ToLowerCaseNames(this ModelBuilder modelBuilder)
>     {
>         _culture = CultureInfo.InvariantCulture;
> 
>         SetNames(modelBuilder, NamingConvention.LowerCase);
>     }
> 
>     private static string? NameRewriter(this string name, NamingConvention naming)
>     {
>         if (string.IsNullOrEmpty(name)) return name;
> 
>         return naming == NamingConvention.SnakeCase
>             ? SnakeCaseNameRewriter(name)
>             : LowerCaseNameRewriter(name);
>     }
> 
>     private enum NamingConvention
>     {
>         SnakeCase,
>         LowerCase,
>     }
> 
>     private static void SetNames(ModelBuilder modelBuilder, NamingConvention naming)
>     {
>         _culture = CultureInfo.InvariantCulture;
> 
>         foreach (var entity in modelBuilder.Model.GetEntityTypes())
>         {
>             entity.SetViewName(entity.GetViewName()?.NameRewriter(naming));
>             entity.SetSchema(entity.GetSchema()?.NameRewriter(naming));
>             entity.SetTableName(entity.GetTableName()?.NameRewriter(naming));
> 
>             foreach (var property in entity!.GetProperties())
>             {
>                 property.SetColumnName(property.GetColumnName()?.NameRewriter(naming));
>             }
> 
>             foreach (var key in entity.GetKeys())
>             {
>                 key.SetName(key.GetName()?.NameRewriter(naming));
>             }
> 
>             foreach (var key in entity.GetForeignKeys())
>             {
>                 key.SetConstraintName(key.GetConstraintName()?.NameRewriter(naming));
>             }
> 
>             foreach (var index in entity.GetIndexes())
>             {
>                 index.SetDatabaseName(index.GetDatabaseName()?.NameRewriter(naming));
>             }
>         }
>     }
> 
>     private static string? LowerCaseNameRewriter(string name)
>         => name.ToLower(_culture);
> 
>     // https://github.com/efcore/EFCore.NamingConventions/blob/main/EFCore.NamingConventions/Internal/SnakeCaseNameRewriter.cs
>     private static string SnakeCaseNameRewriter(string name)
>     {
>         var builder = new StringBuilder(name.Length + Math.Min(2, name.Length / 5));
>         var previousCategory = default(UnicodeCategory?);
> 
>         for (var currentIndex = 0; currentIndex < name.Length; currentIndex++)
>         {
>             var currentChar = name[currentIndex];
>             if (currentChar == '_')
>             {
>                 builder.Append('_');
>                 previousCategory = null;
>                 continue;
>             }
> 
>             var currentCategory = char.GetUnicodeCategory(currentChar);
>             switch (currentCategory)
>             {
>                 case UnicodeCategory.UppercaseLetter:
>                 case UnicodeCategory.TitlecaseLetter:
>                     if (previousCategory == UnicodeCategory.SpaceSeparator ||
>                         previousCategory == UnicodeCategory.LowercaseLetter ||
>                         previousCategory != UnicodeCategory.DecimalDigitNumber &&
>                         previousCategory != null &&
>                         currentIndex > 0 &&
>                         currentIndex + 1 < name.Length &&
>                         char.IsLower(name[currentIndex + 1]))
>                     {
>                         builder.Append('_');
>                     }
> 
>                     currentChar = char.ToLower(currentChar, _culture);
>                     break;
> 
>                 case UnicodeCategory.LowercaseLetter:
>                 case UnicodeCategory.DecimalDigitNumber:
>                     if (previousCategory == UnicodeCategory.SpaceSeparator)
>                     {
>                         builder.Append('_');
>                     }
>                     break;
> 
>                 default:
>                     if (previousCategory != null)
>                     {
>                         previousCategory = UnicodeCategory.SpaceSeparator;
>                     }
>                     continue;
>             }
> 
>             builder.Append(currentChar);
>             previousCategory = currentCategory;
>         }
> 
>         return builder.ToString().ToLower(_culture);
>     }
> }
> ```
> 
> In your OnModelCreating add the call!
> 
> ```cs
> protected override void OnModelCreating(ModelBuilder builder)
> { 
>     builder.ToSnakeCaseNames();
>    // or 
>    //    builder.ToLowerCaseNames();
> }
> ```
> 
> With this you should be able to update your application!
> 
> FYI: Remove the dependency on EFCore.NamingConventions temporarily.

If this works fine, what's the point of EFCore.NamingConventions?

**Comment by HaikAsatryan** [comment]
@mhosman the whole point is that this NuGet packages ha 28mln+ downloads and is battle tested.
@roji Thank you for all your efforts on this project, it's greatly appreciated! 🙏 While we completely understand the workload and the time needed to ensure quality, it’s a bit challenging for production environments to rely on custom builds as a temporary measure. We’re looking forward to the release when you have the chance. Thanks again for all your hard work!

**Comment by fergalmoran** [comment]
> I sometimes wonder why this isn't the default for the npgsql driver given [even Postgres themselves suggest to not use upper case names](https://wiki.postgresql.org/wiki/Don't_Do_This#Don.27t_use_upper_case_table_or_column_names).

I'm not sure if it should be the _default_ in the ngpsql provider (I'm guessing there are nuances here that we're not aware of?) but it would be great if it were at least an option in the ngpsql provider rather than a separate package.

**Comment by roji** [comment]
> but it would be great if it were at least an option in the ngpsql provider rather than a separate package.

Why exactly?

**Comment by pdevito3** [comment]
> Why exactly?

For me, the current EF defaults make column names that don't play nice with Postgres. The main reason I use this lib is for snake case since that's the generally accepted standard and doesn't complicate your sql (eg `select "FirstName" from patients` vs select first_name from patients`).

Currently there's no way to do this OOTB (that I know of without this lib). I definitely get by with this lib but having something baked in does make sense to me.

**Comment by roji** [comment]
I'm not really asking about whether snake case should be the default or not, but rather why this functionality needs to in the Npgsql provider (as you wrote), rather than as an external library. What's wrong with having an additional library for this?

**Comment by pdevito3** [comment]
Not necessarily saying it should be the default, just that the default in EF doesn't play great with [Postgres](https://www.postgresql.org/docs/7.0/syntax525.htm) so many people will probably need to reach for this or do it manually to get around the default naming (and inevitably miss columns).

As far as a baked in vs using the lib, not having any baked in solution for what is a common need with the pg  provider isn't ideal, but we obviously get by with an external provider save for the occasional hurdle like this dependency upgrade.

More of a nice to have in my opinion, as long as we have some way to do it.

**Comment by ProTip** [comment]
I could argue it should be the default for the EF integration. EF generates table names automatically, whereas `npgsql` itself does not?

Maybe this could be rolled into the `efcore.pg` project so it can piggy-back that build and lifecycle.

**Comment by bit365** [comment]
You can use https://github.com/bit365/efcore-naming-converter for .NET 9.0 conversion. We will continue to improve this library in the future. Thank you for your support.

**Comment by mehrdad-tat** [comment]
> You can use https://github.com/bit365/efcore-naming-converter for .NET 9.0 conversion. We will continue to improve this library in the future. Thank you for your support.

The main point is that having a NuGet registry available for GitHub Actions is a thing, genius! 😁

**Comment by neozhu** [comment]
Thank you for your hard work on this project! Any chance we could see a .NET 9 compatible version soon? Even a pre-release would help a lot as many of us are currently blocked.
Appreciate all the effort!

**Comment by fergalmoran** [comment]
> > but it would be great if it were at least an option in the ngpsql provider rather than a separate package.
> 
> Why exactly?

I think it's a common enough use case that it could be part of the standard provider, also there would be one less package to update so wouldn't us snake_casers be able to update to the next .NET version sooner? ngpsql was updated to .NET 9 pretty quickly? I know we can build it ourselves but this doesn't play well with CI servers.

Again - thanks for all your hard work, it's genuinely appreciated.

**Comment by mhosman** [comment]
Agree with @fergalmoran

**Comment by jamiehankins** [comment]
It's a shame that [`System.Text.Json.JsonSeparatorNamingPolicy`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/Common/JsonSeparatorNamingPolicy.cs) is `internal`. There is a perfectly good implementation there for converting between casings.

**Comment by roji** [comment]
Hi all, I've just release 9.0.0 to nuget.org. I haven't had any time to work on bugfixes/improvements for now, but at least there's a version up that should unblock you.
